lint against a template with required behaves oddly

Output of `helm version`: 2.11.0

Since #4221 was merged I started playing around with removing some of my internal `required` checks in CI and I noticed an interesting behaviour which I'm unsure if it is a bug or not.

Looks to be due to it is allowing `nil` through unhindered which does not work great when doing something like `required "API Token required" .token | b64enc | quote`

Running `helm lint` against a template with `required` which then feeds into ` b64enc`:

```
helm lint
[INFO] Missing required value: API Token required
==> Linting .
[ERROR] templates/: render error in "templates/secrets.yaml": template: templates/secrets.yaml:9:77: executing "templates/secrets.yaml" at <b64enc>: invalid value; expected string

Error: 1 chart(s) linted, 1 chart(s) failed
```

When I placed an empty string in values.yaml or did an inline `default` - note, it has to be done with `()` otherwise `required` throws about too many args or similar

```
helm lint
[INFO] Missing required value: API Token required
==> Linting .

1 chart(s) linted, no failures
```

The other oddness is how the info message happens prior to `==> Linting`

--- a/pkg/engine/engine.go
+++ b/pkg/engine/engine.go
@@ -159,7 +159,7 @@ func (e *Engine) alterFuncMap(t *template.Template, referenceTpls map[string]ren
 			if e.LintMode {
 				// Don't fail on missing required values when linting
 				log.Printf("[INFO] Missing required value: %s", warn)
-				return val, nil
+				return "", nil
 			}
 			// Convert nil to "" in case required is piped into other functions
 			return "", fmt.Errorf(warn)

--- a/pkg/engine/engine.go
+++ b/pkg/engine/engine.go
@@ -161,7 +161,8 @@ func (e *Engine) alterFuncMap(t *template.Template, referenceTpls map[string]ren
 				log.Printf("[INFO] Missing required value: %s", warn)
 				return val, nil
 			}
-			return val, fmt.Errorf(warn)
+			// Convert nil to "" in case required is piped into other functions
+			return "", fmt.Errorf(warn)
 		} else if _, ok := val.(string); ok {
 			if val == "" {
 				if e.LintMode {

--- a/pkg/engine/engine_test.go
+++ b/pkg/engine/engine_test.go
@@ -466,12 +466,37 @@ func TestAlterFuncMap(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-
 	expectStr := "All your base are belong to us"
 	if gotStr := outReq["conan/templates/quote"]; gotStr != expectStr {
 		t.Errorf("Expected %q, got %q (%v)", expectStr, gotStr, outReq)
 	}
 	expectNum := "All 2 of them!"
+	if gotNum := outReq["conan/templates/bases"]; gotNum != expectNum {
+		t.Errorf("Expected %q, got %q (%v)", expectNum, gotNum, outReq)
+	}
+
+	// test required without passing in needed values with lint mode on
+	// verifies lint replaces required with an empty string (should not fail)
+	lintValues := chartutil.Values{
+		"Values": chartutil.Values{
+			"who": "us",
+		},
+		"Chart": reqChart.Metadata,
+		"Release": chartutil.Values{
+			"Name": "That 90s meme",
+		},
+	}
+	e := New()
+	e.LintMode = true
+	outReq, err = e.Render(reqChart, lintValues)
+	if err != nil {
+		t.Fatal(err)
+	}
+	expectStr = "All your base are belong to us"
+	if gotStr := outReq["conan/templates/quote"]; gotStr != expectStr {
+		t.Errorf("Expected %q, got %q (%v)", expectStr, gotStr, outReq)
+	}
+	expectNum = "All  of them!"
 	if gotNum := outReq["conan/templates/bases"]; gotNum != expectNum {
 		t.Errorf("Expected %q, got %q (%v)", expectNum, gotNum, outReq)
 	}

